{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "91d261ce",
   "metadata": {},
   "source": [
    "**Purpose**\n",
    "\n",
    "Read in codon record table and synonymous variant record table, filter both to keep records with full sequence contexts, records for codons with >1 degeneracy, and records without STOP codons in the sequence context"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d81c3fb3-0800-462c-9820-e9a7702a307f",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15913de1-e6a6-4d8d-84b2-1a7bd8c0a8cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import importlib.util\n",
    "from textwrap import wrap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3c06b75d-4ddd-42f3-a9b5-888806f4c85e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e515440b-9126-49f9-8b5a-dac0dce24bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "ccv_spec = importlib.util.spec_from_file_location(\"codon_context_variables\", \n",
    "                                                  \"../codon_context_variables.py\")\n",
    "ccv = importlib.util.module_from_spec(ccv_spec)\n",
    "ccv_spec.loader.exec_module(ccv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c04c8e5-4239-4a6b-a851-b8f17463f679",
   "metadata": {},
   "source": [
    "## Files - read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1244db4d-6e9d-45ae-bcb5-0de45957fa97",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_codon_cp3_filename = \"../../data/0_data_processing/rna_stability_exports/RNAStability_v10.5.1_hg38_filterDups_noOverlaps_CP3.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fe7c300c-219c-4837-9a02-5d4ce27309ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_anno_filename = \"../../data/0_data_processing/transcripts/RNAStability_v10.5.1_transcript_entrez_summary.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4aeae453-cd54-404a-b6b3-16fcc8bf7794",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_variant_syn_filename = \"../../data/0_data_processing/rna_stability_exports/RNAStability_v10.5.1_hg38_filterDups_noOverlaps_synonymous.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57d29926-e52b-4ebc-9daf-817bef25fb5c",
   "metadata": {},
   "source": [
    "## Files - write"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7ef5d6ef-e5d9-4ff7-bc08-f1575c9030f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_codon_cp3_noedge_syn_noSTOP_filename = \\\n",
    "    \"../../data/0_data_processing/rna_stability_exports/RNAStability_v10.5.1_hg38_filterDups_noOverlaps_noEdge_wSyn_noSTOP_CP3.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "36739f4d-bafd-41f2-bb30-77df2bec4c97",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_codon_cp3_noedge_syn_noSTOP_nostruct_filename = \\\n",
    "    \"../../data/0_data_processing/rna_stability_exports/RNAStability_v10.5.1_hg38_filterDups_noOverlaps_noEdge_wSyn_noSTOP_noStruct_CP3.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3d5565d6-bde2-472a-84ca-6daa50816fd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_anno_selected_0a_filename = \"../../data/0_data_processing/transcripts/RNAStability_v10.5.1_transcript_entrez_selected_0a.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66470228-28b5-4aa1-b777-14b52d88ee8c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "root = \"../../data/0_data_processing/rna_stability_exports/byAminoAcid_sub/RNAStability_v10.5.1_hg38_filterDups_noOverlaps_synonymous_noEdge_noSTOP_CP3_\"\n",
    "rnas_variant_syn_out = {}\n",
    "for amin in ccv.aa_sub_with_syn_noSTOP :\n",
    "    rnas_variant_syn_out[amin] = {\"seq_filename\": (root+\"seqCol_AminoAcid\"+amin+\".tsv\"),\n",
    "                                  \"nonseq_filename\": (root+\"nonseqCol_AminoAcid\"+amin+\".tsv\")}\n",
    "rnas_variant_syn_out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4782cdd4-2a97-4d3c-b7dc-ec59722b7f0f",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "58afe03d-8cea-4e90-896b-63da45abf8ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_contexts_for_codons (r, codons_omit=ccv.codons_STOP) :\n",
    "    window_pos = r[\"WindowPosition\"]\n",
    "    codon_pos = r[\"CodonPosition\"]\n",
    "    x = r[\"REF_Sequence\"]\n",
    "    \n",
    "    if (window_pos%3 == 0 and codon_pos == 3):\n",
    "        codons_all = wrap(x,3)\n",
    "    elif (window_pos == 51 and codon_pos == 2):\n",
    "        codons_all = wrap(x[1:],3)\n",
    "    elif (window_pos == 51 and codon_pos == 1) :\n",
    "        codons_all = wrap(x[2:],3)\n",
    "    else :\n",
    "        print(\"Not an expected configuration.\")\n",
    "        \n",
    "    if len(set(codons_all) & set(codons_omit)) > 0 :\n",
    "        return False\n",
    "    else :\n",
    "        return True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42255f3b-4fc6-4299-9b7a-b55800e1bbf9",
   "metadata": {},
   "source": [
    "# Filter codon table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "099063da-a51b-4bbc-bfbc-7eb325a74c71",
   "metadata": {},
   "source": [
    "Read in main codon data frame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6d9d2bc9-5a1f-46d6-9fb8-eaac5eef382c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>NM_ID</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>symbol</th>\n",
       "      <th>Sense</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_Stop</th>\n",
       "      <th>CDS_Length</th>\n",
       "      <th>Trans_Length</th>\n",
       "      <th>Trans_POS</th>\n",
       "      <th>PosRelative2Start</th>\n",
       "      <th>PosRelative2Stop</th>\n",
       "      <th>PosRelative2StartPct</th>\n",
       "      <th>WindowPosition</th>\n",
       "      <th>REF_Base</th>\n",
       "      <th>PrecedingBase</th>\n",
       "      <th>TrailingBase</th>\n",
       "      <th>PrecedingBases</th>\n",
       "      <th>TrailingBases</th>\n",
       "      <th>TribaseContext</th>\n",
       "      <th>CodonPosition</th>\n",
       "      <th>REF_Codon</th>\n",
       "      <th>PrecedingCodon</th>\n",
       "      <th>TrailingCodon</th>\n",
       "      <th>PrecedingBicodon</th>\n",
       "      <th>TrailingBicodon</th>\n",
       "      <th>TricodonContext</th>\n",
       "      <th>REF_AminoAcid</th>\n",
       "      <th>REF_AminoAcid_sub</th>\n",
       "      <th>REF_Sequence</th>\n",
       "      <th>REF_mfeValue</th>\n",
       "      <th>REF_cfeValue</th>\n",
       "      <th>REF_meafeValue</th>\n",
       "      <th>REF_efeValue</th>\n",
       "      <th>REF_cdValue</th>\n",
       "      <th>REF_endValue</th>\n",
       "      <th>REF_meaValue</th>\n",
       "      <th>REF_freqMfeEnsemble</th>\n",
       "      <th>REF_mfeStructure</th>\n",
       "      <th>REF_cfeStructure</th>\n",
       "      <th>REF_meafeStructure</th>\n",
       "      <th>REF_efeStructure</th>\n",
       "      <th>REF_BondMFE</th>\n",
       "      <th>REF_LocalMFEStructure</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>149977727</td>\n",
       "      <td>T</td>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>1836</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>343</td>\n",
       "      <td>75</td>\n",
       "      <td>2146</td>\n",
       "      <td>0.034</td>\n",
       "      <td>51</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ATC</td>\n",
       "      <td>GGG</td>\n",
       "      <td>CTG</td>\n",
       "      <td>3</td>\n",
       "      <td>TCT</td>\n",
       "      <td>CCA</td>\n",
       "      <td>GGG</td>\n",
       "      <td>CCATCT</td>\n",
       "      <td>TCTGGG</td>\n",
       "      <td>CCATCTGGG</td>\n",
       "      <td>S</td>\n",
       "      <td>S4</td>\n",
       "      <td>CATAACGTTTCACCCAGAGACTCAGCTGAAGGAAATGACAGTTATC...</td>\n",
       "      <td>-18.1</td>\n",
       "      <td>-9.4</td>\n",
       "      <td>-10.2</td>\n",
       "      <td>-20.33</td>\n",
       "      <td>22.62</td>\n",
       "      <td>29.89</td>\n",
       "      <td>63.62</td>\n",
       "      <td>0.026939</td>\n",
       "      <td>......((((((((((((.....(((((..........)))))......</td>\n",
       "      <td>......(((((.....)))))...((.((((..................</td>\n",
       "      <td>......(((((.....)))))...((.((((.............((...</td>\n",
       "      <td>......(((((.,,,,||}}}..,(((({{(....,..|}}},.{{...</td>\n",
       "      <td>)</td>\n",
       "      <td>.))))))..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>149981141</td>\n",
       "      <td>T</td>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>1836</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>1816</td>\n",
       "      <td>1548</td>\n",
       "      <td>673</td>\n",
       "      <td>0.697</td>\n",
       "      <td>51</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>GTT</td>\n",
       "      <td>GTT</td>\n",
       "      <td>TTG</td>\n",
       "      <td>3</td>\n",
       "      <td>TTT</td>\n",
       "      <td>TGG</td>\n",
       "      <td>GTT</td>\n",
       "      <td>TGGTTT</td>\n",
       "      <td>TTTGTT</td>\n",
       "      <td>TGGTTTGTT</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>GATCTTCCCAAAATGTGGAGTATTAGTAGAATGGATACAGTTATCT...</td>\n",
       "      <td>-17.7</td>\n",
       "      <td>-11.3</td>\n",
       "      <td>-16.6</td>\n",
       "      <td>-20.70</td>\n",
       "      <td>17.00</td>\n",
       "      <td>25.15</td>\n",
       "      <td>75.20</td>\n",
       "      <td>0.007678</td>\n",
       "      <td>.......(((.....)))(((((((((((..((((.(((((....(...</td>\n",
       "      <td>..................(((((((((((...(((.(((((........</td>\n",
       "      <td>.......(((.....)))(((((((((((...(((.(((((....(...</td>\n",
       "      <td>......,(({....,}))(((((((((((..,(((.(((((....{...</td>\n",
       "      <td>.</td>\n",
       "      <td>((.....))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>149981291</td>\n",
       "      <td>G</td>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>1836</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>1966</td>\n",
       "      <td>1698</td>\n",
       "      <td>523</td>\n",
       "      <td>0.765</td>\n",
       "      <td>51</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TGT</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TGT</td>\n",
       "      <td>3</td>\n",
       "      <td>GTG</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TCTGTG</td>\n",
       "      <td>GTGTCT</td>\n",
       "      <td>TCTGTGTCT</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>AAGAGTTCACTGCTTGGCTTGGTGGAAGAGTCTGAGGTCTTTGAAT...</td>\n",
       "      <td>-26.9</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-24.2</td>\n",
       "      <td>-30.09</td>\n",
       "      <td>16.50</td>\n",
       "      <td>25.12</td>\n",
       "      <td>74.62</td>\n",
       "      <td>0.005638</td>\n",
       "      <td>.((((((...(((((((((((........(((((((((.((((((....</td>\n",
       "      <td>..(((((...(((((((((((........(((((.(((...........</td>\n",
       "      <td>..(((((...(((((((((((........(((((((((((.((......</td>\n",
       "      <td>.,(((({...(((((((((((,......,(((((((((,({,,......</td>\n",
       "      <td>.</td>\n",
       "      <td>.......).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>149978135</td>\n",
       "      <td>T</td>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>1836</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>751</td>\n",
       "      <td>483</td>\n",
       "      <td>1738</td>\n",
       "      <td>0.218</td>\n",
       "      <td>51</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>CTC</td>\n",
       "      <td>GTG</td>\n",
       "      <td>CTG</td>\n",
       "      <td>3</td>\n",
       "      <td>TCT</td>\n",
       "      <td>ATC</td>\n",
       "      <td>GTG</td>\n",
       "      <td>ATCTCT</td>\n",
       "      <td>TCTGTG</td>\n",
       "      <td>ATCTCTGTG</td>\n",
       "      <td>S</td>\n",
       "      <td>S4</td>\n",
       "      <td>TTTTTTGCCAGCATCATTTATTTTCTCTTGGGTACCTCCCGTCACA...</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-17.8</td>\n",
       "      <td>-18.2</td>\n",
       "      <td>-22.46</td>\n",
       "      <td>17.01</td>\n",
       "      <td>24.82</td>\n",
       "      <td>74.72</td>\n",
       "      <td>0.025643</td>\n",
       "      <td>..((((((((..(((((............(((.....)))..((((...</td>\n",
       "      <td>..((((((((...................(((.....)))..((((...</td>\n",
       "      <td>..((((((((...(((.............(((.....)))..((((...</td>\n",
       "      <td>..{{{(((((,.,{{{,...........,(((.....)))}.((((...</td>\n",
       "      <td>)</td>\n",
       "      <td>...))))((</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>149978075</td>\n",
       "      <td>G</td>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>1836</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>691</td>\n",
       "      <td>423</td>\n",
       "      <td>1798</td>\n",
       "      <td>0.191</td>\n",
       "      <td>51</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TGT</td>\n",
       "      <td>3</td>\n",
       "      <td>CTG</td>\n",
       "      <td>GGT</td>\n",
       "      <td>TAC</td>\n",
       "      <td>GGTCTG</td>\n",
       "      <td>CTGTAC</td>\n",
       "      <td>GGTCTGTAC</td>\n",
       "      <td>L</td>\n",
       "      <td>L4</td>\n",
       "      <td>CAGTCCATTGCTTATTCCCTGCTGGCTGGCCAAGAACCTGTCTATG...</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-14.7</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-21.88</td>\n",
       "      <td>13.43</td>\n",
       "      <td>20.31</td>\n",
       "      <td>80.37</td>\n",
       "      <td>0.065354</td>\n",
       "      <td>...................(((((((.((((((((.....))).))...</td>\n",
       "      <td>...................(((((((..((((.(.......)..))...</td>\n",
       "      <td>...................(((((((.((((((((.....))).))...</td>\n",
       "      <td>...................{((((((.((({((((.....|}}.})...</td>\n",
       "      <td>.</td>\n",
       "      <td>)).......</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHR        POS REF        NM_ID  entrezgene   symbol Sense  CDS_Start  \\\n",
       "0   5  149977727   T  NM_000112.3        1836  SLC26A2     +        269   \n",
       "1   5  149981141   T  NM_000112.3        1836  SLC26A2     +        269   \n",
       "2   5  149981291   G  NM_000112.3        1836  SLC26A2     +        269   \n",
       "3   5  149978135   T  NM_000112.3        1836  SLC26A2     +        269   \n",
       "4   5  149978075   G  NM_000112.3        1836  SLC26A2     +        269   \n",
       "\n",
       "   CDS_Stop  CDS_Length  Trans_Length  Trans_POS  PosRelative2Start  \\\n",
       "0      2488        2220          8082        343                 75   \n",
       "1      2488        2220          8082       1816               1548   \n",
       "2      2488        2220          8082       1966               1698   \n",
       "3      2488        2220          8082        751                483   \n",
       "4      2488        2220          8082        691                423   \n",
       "\n",
       "   PosRelative2Stop  PosRelative2StartPct  WindowPosition REF_Base  \\\n",
       "0              2146                 0.034              51        T   \n",
       "1               673                 0.697              51        T   \n",
       "2               523                 0.765              51        G   \n",
       "3              1738                 0.218              51        T   \n",
       "4              1798                 0.191              51        G   \n",
       "\n",
       "  PrecedingBase TrailingBase PrecedingBases TrailingBases TribaseContext  \\\n",
       "0             C            G            ATC           GGG            CTG   \n",
       "1             T            G            GTT           GTT            TTG   \n",
       "2             T            T            TGT           TCT            TGT   \n",
       "3             C            G            CTC           GTG            CTG   \n",
       "4             T            T            TCT           TAC            TGT   \n",
       "\n",
       "   CodonPosition REF_Codon PrecedingCodon TrailingCodon PrecedingBicodon  \\\n",
       "0              3       TCT            CCA           GGG           CCATCT   \n",
       "1              3       TTT            TGG           GTT           TGGTTT   \n",
       "2              3       GTG            TCT           TCT           TCTGTG   \n",
       "3              3       TCT            ATC           GTG           ATCTCT   \n",
       "4              3       CTG            GGT           TAC           GGTCTG   \n",
       "\n",
       "  TrailingBicodon TricodonContext REF_AminoAcid REF_AminoAcid_sub  \\\n",
       "0          TCTGGG       CCATCTGGG             S                S4   \n",
       "1          TTTGTT       TGGTTTGTT             F                 F   \n",
       "2          GTGTCT       TCTGTGTCT             V                 V   \n",
       "3          TCTGTG       ATCTCTGTG             S                S4   \n",
       "4          CTGTAC       GGTCTGTAC             L                L4   \n",
       "\n",
       "                                        REF_Sequence  REF_mfeValue  \\\n",
       "0  CATAACGTTTCACCCAGAGACTCAGCTGAAGGAAATGACAGTTATC...         -18.1   \n",
       "1  GATCTTCCCAAAATGTGGAGTATTAGTAGAATGGATACAGTTATCT...         -17.7   \n",
       "2  AAGAGTTCACTGCTTGGCTTGGTGGAAGAGTCTGAGGTCTTTGAAT...         -26.9   \n",
       "3  TTTTTTGCCAGCATCATTTATTTTCTCTTGGGTACCTCCCGTCACA...         -20.2   \n",
       "4  CAGTCCATTGCTTATTCCCTGCTGGCTGGCCAAGAACCTGTCTATG...         -20.2   \n",
       "\n",
       "   REF_cfeValue  REF_meafeValue  REF_efeValue  REF_cdValue  REF_endValue  \\\n",
       "0          -9.4           -10.2        -20.33        22.62         29.89   \n",
       "1         -11.3           -16.6        -20.70        17.00         25.15   \n",
       "2         -19.9           -24.2        -30.09        16.50         25.12   \n",
       "3         -17.8           -18.2        -22.46        17.01         24.82   \n",
       "4         -14.7           -20.2        -21.88        13.43         20.31   \n",
       "\n",
       "   REF_meaValue  REF_freqMfeEnsemble  \\\n",
       "0         63.62             0.026939   \n",
       "1         75.20             0.007678   \n",
       "2         74.62             0.005638   \n",
       "3         74.72             0.025643   \n",
       "4         80.37             0.065354   \n",
       "\n",
       "                                    REF_mfeStructure  \\\n",
       "0  ......((((((((((((.....(((((..........)))))......   \n",
       "1  .......(((.....)))(((((((((((..((((.(((((....(...   \n",
       "2  .((((((...(((((((((((........(((((((((.((((((....   \n",
       "3  ..((((((((..(((((............(((.....)))..((((...   \n",
       "4  ...................(((((((.((((((((.....))).))...   \n",
       "\n",
       "                                    REF_cfeStructure  \\\n",
       "0  ......(((((.....)))))...((.((((..................   \n",
       "1  ..................(((((((((((...(((.(((((........   \n",
       "2  ..(((((...(((((((((((........(((((.(((...........   \n",
       "3  ..((((((((...................(((.....)))..((((...   \n",
       "4  ...................(((((((..((((.(.......)..))...   \n",
       "\n",
       "                                  REF_meafeStructure  \\\n",
       "0  ......(((((.....)))))...((.((((.............((...   \n",
       "1  .......(((.....)))(((((((((((...(((.(((((....(...   \n",
       "2  ..(((((...(((((((((((........(((((((((((.((......   \n",
       "3  ..((((((((...(((.............(((.....)))..((((...   \n",
       "4  ...................(((((((.((((((((.....))).))...   \n",
       "\n",
       "                                    REF_efeStructure REF_BondMFE  \\\n",
       "0  ......(((((.,,,,||}}}..,(((({{(....,..|}}},.{{...           )   \n",
       "1  ......,(({....,}))(((((((((((..,(((.(((((....{...           .   \n",
       "2  .,(((({...(((((((((((,......,(((((((((,({,,......           .   \n",
       "3  ..{{{(((((,.,{{{,...........,(((.....)))}.((((...           )   \n",
       "4  ...................{((((((.((({((((.....|}}.})...           .   \n",
       "\n",
       "  REF_LocalMFEStructure  \n",
       "0             .))))))..  \n",
       "1             ((.....))  \n",
       "2             .......).  \n",
       "3             ...))))((  \n",
       "4             )).......  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_df = pd.read_csv(rnas_codon_cp3_filename,\n",
    "                                sep=\"\\t\",\n",
    "                                dtype={\"CHR\":str})\n",
    "rnas_codon_cp3_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3c6338d9-de51-4bcf-8a7b-651057305ae4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHR                       object\n",
       "POS                        int64\n",
       "REF                       object\n",
       "NM_ID                     object\n",
       "entrezgene                 int64\n",
       "symbol                    object\n",
       "Sense                     object\n",
       "CDS_Start                  int64\n",
       "CDS_Stop                   int64\n",
       "CDS_Length                 int64\n",
       "Trans_Length               int64\n",
       "Trans_POS                  int64\n",
       "PosRelative2Start          int64\n",
       "PosRelative2Stop           int64\n",
       "PosRelative2StartPct     float64\n",
       "WindowPosition             int64\n",
       "REF_Base                  object\n",
       "PrecedingBase             object\n",
       "TrailingBase              object\n",
       "PrecedingBases            object\n",
       "TrailingBases             object\n",
       "TribaseContext            object\n",
       "CodonPosition              int64\n",
       "REF_Codon                 object\n",
       "PrecedingCodon            object\n",
       "TrailingCodon             object\n",
       "PrecedingBicodon          object\n",
       "TrailingBicodon           object\n",
       "TricodonContext           object\n",
       "REF_AminoAcid             object\n",
       "REF_AminoAcid_sub         object\n",
       "REF_Sequence              object\n",
       "REF_mfeValue             float64\n",
       "REF_cfeValue             float64\n",
       "REF_meafeValue           float64\n",
       "REF_efeValue             float64\n",
       "REF_cdValue              float64\n",
       "REF_endValue             float64\n",
       "REF_meaValue             float64\n",
       "REF_freqMfeEnsemble      float64\n",
       "REF_mfeStructure          object\n",
       "REF_cfeStructure          object\n",
       "REF_meafeStructure        object\n",
       "REF_efeStructure          object\n",
       "REF_BondMFE               object\n",
       "REF_LocalMFEStructure     object\n",
       "dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_df.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad75eead-afbb-4c2c-91f1-07ebfe8b6e5d",
   "metadata": {},
   "source": [
    "Check style of CHR column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "61ae65df-0250-4429-8ae5-372fac817bc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1     1097936\n",
       "2      793446\n",
       "19     701390\n",
       "11     634475\n",
       "3      621710\n",
       "17     618977\n",
       "12     572109\n",
       "6      550857\n",
       "7      507095\n",
       "5      473065\n",
       "16     452232\n",
       "9      437438\n",
       "4      433575\n",
       "10     422225\n",
       "X      415509\n",
       "15     369945\n",
       "8      359904\n",
       "14     341870\n",
       "20     254792\n",
       "22     221614\n",
       "13     196113\n",
       "18     170845\n",
       "21     105896\n",
       "Y       27519\n",
       "Name: CHR, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_df[\"CHR\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bc1fe4c-d0cf-4871-93c7-bcb5b86ef388",
   "metadata": {},
   "source": [
    "Check that we have the expected unique keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6e5a4622-c5fd-448e-bade-2664648908b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>NM_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <th>69093</th>\n",
       "      <th>G</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">3</th>\n",
       "      <th>52331224</th>\n",
       "      <th>A</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52331230</th>\n",
       "      <th>G</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52331233</th>\n",
       "      <th>A</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52331236</th>\n",
       "      <th>C</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  NM_ID\n",
       "CHR POS      REF       \n",
       "1   69093    G        1\n",
       "3   52331224 A        1\n",
       "    52331230 G        1\n",
       "    52331233 A        1\n",
       "    52331236 C        1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_df.groupby([\"CHR\", \"POS\", \"REF\"]).\\\n",
    "    agg({\"NM_ID\":len}).\\\n",
    "    sort_values(\"NM_ID\", ascending=False).\\\n",
    "    head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "60e8a81f-48e4-4f13-b1f0-00f9e44d361e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10780537, 46)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "85e45ce0-2131-4de3-9fb6-90c6c840ca4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18718"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_df[\"NM_ID\"].unique().size"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce6d313a-9a16-4ab0-b7d8-4f12a2f00548",
   "metadata": {},
   "source": [
    "## Remove near-end variants and single degeneracy codons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c963b76f-cd95-4a3d-966c-755f513ed018",
   "metadata": {},
   "source": [
    "Select codons whose CP3 is removed a bit from the start codon:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9661b954-d13c-48f2-bbc6-12cc2bd106b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_codon_cp3_noedge_df = rnas_codon_cp3_df.query(\"PosRelative2Start > 51 & \" + \n",
    "                                                   \"PosRelative2Stop > 51\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "48993839-008f-428a-9fe8-857e252d7307",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10144879, 46)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_noedge_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c49fa5d-2450-4ac2-b3f1-2725337da97f",
   "metadata": {},
   "source": [
    "Double check that the sequence context is centered on the codon in all cases now:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a1185651-3d5b-4783-ae60-fc8b7d1c301e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51    10144879\n",
       "Name: WindowPosition, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_noedge_df[\"WindowPosition\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbe4a1f1-af76-4f4a-b733-3b0937cff3c3",
   "metadata": {},
   "source": [
    "Grab codons that have synonyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dd552f21-a84b-4b21-a908-00ea5440ed52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['F',\n",
       " 'L',\n",
       " 'I',\n",
       " 'V',\n",
       " 'S',\n",
       " 'P',\n",
       " 'T',\n",
       " 'A',\n",
       " 'Y',\n",
       " 'H',\n",
       " 'Q',\n",
       " 'N',\n",
       " 'K',\n",
       " 'D',\n",
       " 'E',\n",
       " 'C',\n",
       " 'R',\n",
       " 'G']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccv.aa_with_syn_noSTOP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a4182e5d-bc8e-4269-b25e-8138836d50a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9823355, 46)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_noedge_syn_df = rnas_codon_cp3_noedge_df.query(\"REF_AminoAcid in @ccv.aa_with_syn_noSTOP\")\n",
    "rnas_codon_cp3_noedge_syn_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "400beccf-55f5-4aa8-8f2f-989c4a4eaa73",
   "metadata": {},
   "source": [
    "## Remove records with STOP codons appearing within the sequence context"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e975c5b9-4039-48e0-924c-85cdfcf2fffa",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_context_noSTOP = rnas_codon_cp3_noedge_syn_df[\"REF_Sequence\"].apply(filter_contexts_for_codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5afcaacc-6a93-4cb7-a9b8-41fd7d0754ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_codon_cp3_noedge_syn_noSTOP_df = rnas_codon_cp3_noedge_syn_df[codon_context_noSTOP]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e4d16c71-2fb3-4c6c-b27a-95fd862e2ca2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9822251, 46)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_codon_cp3_noedge_syn_noSTOP_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5024c32-e23b-4ba5-ac82-3c9c7baebc6f",
   "metadata": {},
   "source": [
    "## Save filtered files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d9d1b24-7ba3-497d-b662-7e8a77e4f74f",
   "metadata": {},
   "source": [
    "Remove structure fields that won't be often used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5f0c4963-4894-49c6-b554-f8c8b7e7744c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_codon_cp3_noedge_syn_noSTOP_nostruct_df = \\\n",
    "    rnas_codon_cp3_noedge_syn_noSTOP_df.drop(columns=[\"REF_mfeStructure\",\n",
    "                                                      \"REF_cfeStructure\",\n",
    "                                                      \"REF_meafeStructure\",\n",
    "                                                      \"REF_efeStructure\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad876215-223c-43fd-b7e7-d7f2c71c660e",
   "metadata": {},
   "source": [
    "Save filtered files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "bc6da85e-7f58-4756-99df-2ebe9a07fd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_codon_cp3_noedge_syn_noSTOP_df.to_csv(rnas_codon_cp3_noedge_syn_noSTOP_filename,\n",
    "                                           sep=\"\\t\",\n",
    "                                           index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7a9f2a2-28b2-4f2f-a94d-a64478cbb2d6",
   "metadata": {},
   "source": [
    "Save version that doesn't include structural data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "235ca4de-1786-4899-90bc-8b324401026a",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_codon_cp3_noedge_syn_noSTOP_nostruct_df.to_csv(rnas_codon_cp3_noedge_syn_noSTOP_nostruct_filename,\n",
    "                                                    sep=\"\\t\",\n",
    "                                                    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e302f161-1d6f-414e-982e-eed50fb01a1b",
   "metadata": {},
   "source": [
    "# Transcript information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "2b4f155a-8f67-4fea-8391-56dd1751d886",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18704"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts = rnas_codon_cp3_noedge_syn_noSTOP_df[\"NM_ID\"].unique()\n",
    "len(transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9534fe1c-758e-478f-aed5-a72c5155e0f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NM_ID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>Sense</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_Stop</th>\n",
       "      <th>CDS_Length</th>\n",
       "      <th>Trans_Length</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_019089.4</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>100</td>\n",
       "      <td>621</td>\n",
       "      <td>522</td>\n",
       "      <td>4273</td>\n",
       "      <td>54626.0</td>\n",
       "      <td>hes family bHLH transcription factor 2</td>\n",
       "      <td>HES2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_015484.4</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>56</td>\n",
       "      <td>787</td>\n",
       "      <td>732</td>\n",
       "      <td>1777</td>\n",
       "      <td>25949.0</td>\n",
       "      <td>SYF2 pre-mRNA splicing factor</td>\n",
       "      <td>SYF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_015627.2</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>115</td>\n",
       "      <td>1041</td>\n",
       "      <td>927</td>\n",
       "      <td>2935</td>\n",
       "      <td>26119.0</td>\n",
       "      <td>low density lipoprotein receptor adaptor prote...</td>\n",
       "      <td>LDLRAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001011547.2</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>53</td>\n",
       "      <td>2098</td>\n",
       "      <td>2046</td>\n",
       "      <td>3218</td>\n",
       "      <td>200010.0</td>\n",
       "      <td>solute carrier family 5 member 9</td>\n",
       "      <td>SLC5A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_004153.3</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>232</td>\n",
       "      <td>2817</td>\n",
       "      <td>2586</td>\n",
       "      <td>3192</td>\n",
       "      <td>4998.0</td>\n",
       "      <td>origin recognition complex subunit 1</td>\n",
       "      <td>ORC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            NM_ID CHR Sense  CDS_Start  CDS_Stop  CDS_Length  Trans_Length  \\\n",
       "0     NM_019089.4   1     -        100       621         522          4273   \n",
       "1     NM_015484.4   1     -         56       787         732          1777   \n",
       "2     NM_015627.2   1     +        115      1041         927          2935   \n",
       "3  NM_001011547.2   1     +         53      2098        2046          3218   \n",
       "4     NM_004153.3   1     -        232      2817        2586          3192   \n",
       "\n",
       "   entrezgene                                               name   symbol  \n",
       "0     54626.0             hes family bHLH transcription factor 2     HES2  \n",
       "1     25949.0                      SYF2 pre-mRNA splicing factor     SYF2  \n",
       "2     26119.0  low density lipoprotein receptor adaptor prote...  LDLRAP1  \n",
       "3    200010.0                   solute carrier family 5 member 9   SLC5A9  \n",
       "4      4998.0               origin recognition complex subunit 1     ORC1  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_anno_df = pd.read_csv(transcript_anno_filename,\n",
    "                                 sep=\"\\t\")\n",
    "transcript_anno_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "608b00fc-80e7-497c-a40d-21226b5a21d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(45433, 10)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_anno_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0fa52272-e9bd-458f-ad63-28ec58a3a076",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18704, 10)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_anno_selected_df = transcript_anno_df[transcript_anno_df[\"NM_ID\"].isin(transcripts)]\n",
    "transcript_anno_selected_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76d539b-67bf-4dc9-9f5b-daf1264b61f9",
   "metadata": {},
   "source": [
    "## Save filtered file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "112fd76f-fb4a-4064-b902-c13b435080b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_anno_selected_df.to_csv(transcript_anno_selected_0a_filename,\n",
    "                                   sep=\"\\t\",\n",
    "                                   index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89e51b7c-5601-4c1d-8456-e3ecf7e940d6",
   "metadata": {},
   "source": [
    "# Filter variant table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f72bfbe-f31c-4b95-8703-a1443b7e2548",
   "metadata": {},
   "source": [
    "Read in synonymous variant table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f6dd877b-774b-469b-b944-e9a95875fe7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NM_ID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Sense</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_Stop</th>\n",
       "      <th>CDS_Length</th>\n",
       "      <th>Trans_Length</th>\n",
       "      <th>Trans_POS</th>\n",
       "      <th>REF_Base</th>\n",
       "      <th>ALT_Base</th>\n",
       "      <th>WindowPosition</th>\n",
       "      <th>REF_Sequence</th>\n",
       "      <th>ALT_Sequence</th>\n",
       "      <th>deltaMFE</th>\n",
       "      <th>deltaCFE</th>\n",
       "      <th>deltaMEAFE</th>\n",
       "      <th>deltaEFE</th>\n",
       "      <th>MFEED</th>\n",
       "      <th>CFEED</th>\n",
       "      <th>MEAED</th>\n",
       "      <th>EFEED</th>\n",
       "      <th>deltaCD</th>\n",
       "      <th>deltaEND</th>\n",
       "      <th>REF_mfeValue</th>\n",
       "      <th>REF_cfeValue</th>\n",
       "      <th>REF_meafeValue</th>\n",
       "      <th>REF_efeValue</th>\n",
       "      <th>REF_cdValue</th>\n",
       "      <th>REF_endValue</th>\n",
       "      <th>REF_meaValue</th>\n",
       "      <th>REF_freqMfeEnsemble</th>\n",
       "      <th>ALT_mfeValue</th>\n",
       "      <th>ALT_cfeValue</th>\n",
       "      <th>ALT_meafeValue</th>\n",
       "      <th>ALT_efeValue</th>\n",
       "      <th>ALT_cdValue</th>\n",
       "      <th>ALT_endValue</th>\n",
       "      <th>ALT_meaValue</th>\n",
       "      <th>ALT_freqMfeEnsemble</th>\n",
       "      <th>REF_mfeStructure</th>\n",
       "      <th>REF_cfeStructure</th>\n",
       "      <th>REF_meafeStructure</th>\n",
       "      <th>REF_efeStructure</th>\n",
       "      <th>ALT_mfeStructure</th>\n",
       "      <th>ALT_cfeStructure</th>\n",
       "      <th>ALT_meafeStructure</th>\n",
       "      <th>ALT_efeStructure</th>\n",
       "      <th>GENE</th>\n",
       "      <th>HGVS</th>\n",
       "      <th>LOC_IN_GENE</th>\n",
       "      <th>EFFECT</th>\n",
       "      <th>DIST_TO_CODING_REGION</th>\n",
       "      <th>DIST_TO_SPLICE_SITE</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>gnomAD3_WG_AC</th>\n",
       "      <th>gnomAD3_WG_AN</th>\n",
       "      <th>gnomAD3_WG_AF</th>\n",
       "      <th>gnomAD3_WG_nhomalt</th>\n",
       "      <th>gnomAD3_Coverage</th>\n",
       "      <th>gnomAD3_Quality</th>\n",
       "      <th>gnomAD3_WG_InbreedingCoeff</th>\n",
       "      <th>gnomAD3_CVGE_OVER_20</th>\n",
       "      <th>gnomAD2_EX_AC</th>\n",
       "      <th>gnomAD2_EX_AN</th>\n",
       "      <th>gnomAD2_EX_AF</th>\n",
       "      <th>gnomAD2_EX_nhomalt</th>\n",
       "      <th>gnomAD2_Coverage</th>\n",
       "      <th>gnomAD2_Quality</th>\n",
       "      <th>gnomAD2_EX_InbreedingCoeff</th>\n",
       "      <th>gnomAD2_EX_rf_tp_probability</th>\n",
       "      <th>gnomAD2_CVGE_OVER_20</th>\n",
       "      <th>gnomAD3_WG_vep</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>NM_ID_noVersion</th>\n",
       "      <th>is_MANE</th>\n",
       "      <th>is_LongestCDS</th>\n",
       "      <th>is_LongestTrans</th>\n",
       "      <th>snpeffTYPE</th>\n",
       "      <th>PosRelative2Start</th>\n",
       "      <th>PosRelative2Stop</th>\n",
       "      <th>PosRelative2StartPct</th>\n",
       "      <th>PrecedingBase</th>\n",
       "      <th>TrailingBase</th>\n",
       "      <th>PrecedingBases</th>\n",
       "      <th>TrailingBases</th>\n",
       "      <th>SNVContext</th>\n",
       "      <th>TribaseContext</th>\n",
       "      <th>REF_BondMFE</th>\n",
       "      <th>ALT_BondMFE</th>\n",
       "      <th>REF_LocalMFEStructure</th>\n",
       "      <th>ALT_LocalMFEStructure</th>\n",
       "      <th>CodonPosition</th>\n",
       "      <th>REF_Codon</th>\n",
       "      <th>ALT_Codon</th>\n",
       "      <th>PrecedingCodon</th>\n",
       "      <th>TrailingCodon</th>\n",
       "      <th>PrecedingBicodon</th>\n",
       "      <th>TrailingBicodon</th>\n",
       "      <th>TricodonContext</th>\n",
       "      <th>REF_AminoAcid</th>\n",
       "      <th>ALT_AminoAcid</th>\n",
       "      <th>REF_AminoAcid_sub</th>\n",
       "      <th>ALT_AminoAcid_sub</th>\n",
       "      <th>manualTYPE</th>\n",
       "      <th>DupAlignment_select</th>\n",
       "      <th>maxGeneCount</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>5</td>\n",
       "      <td>149977727</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>343</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>51</td>\n",
       "      <td>CATAACGTTTCACCCAGAGACTCAGCTGAAGGAAATGACAGTTATC...</td>\n",
       "      <td>CATAACGTTTCACCCAGAGACTCAGCTGAAGGAAATGACAGTTATC...</td>\n",
       "      <td>0.9</td>\n",
       "      <td>5.70</td>\n",
       "      <td>3.9</td>\n",
       "      <td>0.60</td>\n",
       "      <td>40</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>9.908430</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.19</td>\n",
       "      <td>-18.1</td>\n",
       "      <td>-9.4</td>\n",
       "      <td>-10.2</td>\n",
       "      <td>-20.33</td>\n",
       "      <td>22.62</td>\n",
       "      <td>29.89</td>\n",
       "      <td>63.62</td>\n",
       "      <td>0.026939</td>\n",
       "      <td>-17.2</td>\n",
       "      <td>-3.70</td>\n",
       "      <td>-6.3</td>\n",
       "      <td>-19.73</td>\n",
       "      <td>22.96</td>\n",
       "      <td>30.08</td>\n",
       "      <td>62.60</td>\n",
       "      <td>0.016616</td>\n",
       "      <td>......((((((((((((.....(((((..........)))))......</td>\n",
       "      <td>......(((((.....)))))...((.((((..................</td>\n",
       "      <td>......(((((.....)))))...((.((((.............((...</td>\n",
       "      <td>......(((((.,,,,||}}}..,(((({{(....,..|}}},.{{...</td>\n",
       "      <td>.............(((((.......((((.((((((....))).))...</td>\n",
       "      <td>......(((((.....)))))............................</td>\n",
       "      <td>......(((((.....)))))...((.((((((...........))...</td>\n",
       "      <td>......((({{..,,,||}}}...{(((({({,{,{....},}.||...</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112.3(SLC26A2):c.75T&gt;A:p.Ser25Ser</td>\n",
       "      <td>coding_sequence</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.973474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.99905</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1836</td>\n",
       "      <td>solute carrier family 26 member 2</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>75</td>\n",
       "      <td>2146</td>\n",
       "      <td>0.034</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ATC</td>\n",
       "      <td>GGG</td>\n",
       "      <td>T&gt;A</td>\n",
       "      <td>CTG</td>\n",
       "      <td>)</td>\n",
       "      <td>)</td>\n",
       "      <td>.))))))..</td>\n",
       "      <td>.))))....</td>\n",
       "      <td>3</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TCA</td>\n",
       "      <td>CCA</td>\n",
       "      <td>GGG</td>\n",
       "      <td>CCATCT</td>\n",
       "      <td>TCTGGG</td>\n",
       "      <td>CCATCTGGG</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S4</td>\n",
       "      <td>S4</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>5</td>\n",
       "      <td>149981141</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>1816</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>51</td>\n",
       "      <td>GATCTTCCCAAAATGTGGAGTATTAGTAGAATGGATACAGTTATCT...</td>\n",
       "      <td>GATCTTCCCAAAATGTGGAGTATTAGTAGAATGGATACAGTTATCT...</td>\n",
       "      <td>-2.2</td>\n",
       "      <td>-6.30</td>\n",
       "      <td>-3.3</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>38</td>\n",
       "      <td>46</td>\n",
       "      <td>40</td>\n",
       "      <td>22.772300</td>\n",
       "      <td>1.68</td>\n",
       "      <td>2.89</td>\n",
       "      <td>-17.7</td>\n",
       "      <td>-11.3</td>\n",
       "      <td>-16.6</td>\n",
       "      <td>-20.70</td>\n",
       "      <td>17.00</td>\n",
       "      <td>25.15</td>\n",
       "      <td>75.20</td>\n",
       "      <td>0.007678</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-17.60</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-21.98</td>\n",
       "      <td>18.68</td>\n",
       "      <td>28.04</td>\n",
       "      <td>72.11</td>\n",
       "      <td>0.034310</td>\n",
       "      <td>.......(((.....)))(((((((((((..((((.(((((....(...</td>\n",
       "      <td>..................(((((((((((...(((.(((((........</td>\n",
       "      <td>.......(((.....)))(((((((((((...(((.(((((....(...</td>\n",
       "      <td>......,(({....,}))(((((((((((..,(((.(((((....{...</td>\n",
       "      <td>.............(((((((.((..(((((((((((.(((....))...</td>\n",
       "      <td>.............(((((((.....(((((((((((.(((....))...</td>\n",
       "      <td>.............(((((((.((..(((((((((((.(((....))...</td>\n",
       "      <td>........,....{((((((.({,,(((((((((({.(((....))...</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112.3(SLC26A2):c.1548T&gt;C:p.Phe516Phe</td>\n",
       "      <td>coding_sequence</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.965803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.99800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1836</td>\n",
       "      <td>solute carrier family 26 member 2</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1548</td>\n",
       "      <td>673</td>\n",
       "      <td>0.697</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>GTT</td>\n",
       "      <td>GTT</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>TTG</td>\n",
       "      <td>.</td>\n",
       "      <td>)</td>\n",
       "      <td>((.....))</td>\n",
       "      <td>))))))).)</td>\n",
       "      <td>3</td>\n",
       "      <td>TTT</td>\n",
       "      <td>TTC</td>\n",
       "      <td>TGG</td>\n",
       "      <td>GTT</td>\n",
       "      <td>TGGTTT</td>\n",
       "      <td>TTTGTT</td>\n",
       "      <td>TGGTTTGTT</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>5</td>\n",
       "      <td>149981291</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>1966</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>51</td>\n",
       "      <td>AAGAGTTCACTGCTTGGCTTGGTGGAAGAGTCTGAGGTCTTTGAAT...</td>\n",
       "      <td>AAGAGTTCACTGCTTGGCTTGGTGGAAGAGTCTGAGGTCTTTGAAT...</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>16</td>\n",
       "      <td>0.395924</td>\n",
       "      <td>-0.84</td>\n",
       "      <td>-1.82</td>\n",
       "      <td>-26.9</td>\n",
       "      <td>-19.9</td>\n",
       "      <td>-24.2</td>\n",
       "      <td>-30.09</td>\n",
       "      <td>16.50</td>\n",
       "      <td>25.12</td>\n",
       "      <td>74.62</td>\n",
       "      <td>0.005638</td>\n",
       "      <td>-27.3</td>\n",
       "      <td>-19.90</td>\n",
       "      <td>-26.7</td>\n",
       "      <td>-30.29</td>\n",
       "      <td>15.66</td>\n",
       "      <td>23.30</td>\n",
       "      <td>76.75</td>\n",
       "      <td>0.007827</td>\n",
       "      <td>.((((((...(((((((((((........(((((((((.((((((....</td>\n",
       "      <td>..(((((...(((((((((((........(((((.(((...........</td>\n",
       "      <td>..(((((...(((((((((((........(((((((((((.((......</td>\n",
       "      <td>.,(((({...(((((((((((,......,(((((((((,({,,......</td>\n",
       "      <td>.((((((...(((((((((((........(((((((((((.........</td>\n",
       "      <td>..(((((...(((((((((((........(((((.(((...........</td>\n",
       "      <td>..(((((...(((((((((((........(((((((((((.........</td>\n",
       "      <td>.,(((({...(((((((((((.......,(((((((((,{,,,......</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112.3(SLC26A2):c.1698G&gt;A:p.Val566Val</td>\n",
       "      <td>coding_sequence</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.954799</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.99916</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1836</td>\n",
       "      <td>solute carrier family 26 member 2</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1698</td>\n",
       "      <td>523</td>\n",
       "      <td>0.765</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TGT</td>\n",
       "      <td>TCT</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>TGT</td>\n",
       "      <td>.</td>\n",
       "      <td>(</td>\n",
       "      <td>.......).</td>\n",
       "      <td>((((.....</td>\n",
       "      <td>3</td>\n",
       "      <td>GTG</td>\n",
       "      <td>GTA</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TCTGTG</td>\n",
       "      <td>GTGTCT</td>\n",
       "      <td>TCTGTGTCT</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>5</td>\n",
       "      <td>149978135</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>751</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>51</td>\n",
       "      <td>TTTTTTGCCAGCATCATTTATTTTCTCTTGGGTACCTCCCGTCACA...</td>\n",
       "      <td>TTTTTTGCCAGCATCATTTATTTTCTCTTGGGTACCTCCCGTCACA...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>12.48</td>\n",
       "      <td>0.7</td>\n",
       "      <td>0.35</td>\n",
       "      <td>68</td>\n",
       "      <td>36</td>\n",
       "      <td>4</td>\n",
       "      <td>5.356520</td>\n",
       "      <td>2.78</td>\n",
       "      <td>3.67</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-17.8</td>\n",
       "      <td>-18.2</td>\n",
       "      <td>-22.46</td>\n",
       "      <td>17.01</td>\n",
       "      <td>24.82</td>\n",
       "      <td>74.72</td>\n",
       "      <td>0.025643</td>\n",
       "      <td>-20.0</td>\n",
       "      <td>-5.32</td>\n",
       "      <td>-17.5</td>\n",
       "      <td>-22.11</td>\n",
       "      <td>19.79</td>\n",
       "      <td>28.49</td>\n",
       "      <td>68.23</td>\n",
       "      <td>0.032343</td>\n",
       "      <td>..((((((((..(((((............(((.....)))..((((...</td>\n",
       "      <td>..((((((((...................(((.....)))..((((...</td>\n",
       "      <td>..((((((((...(((.............(((.....)))..((((...</td>\n",
       "      <td>..{{{(((((,.,{{{,...........,(((.....)))}.((((...</td>\n",
       "      <td>........((((.................(((.....)))(((.((...</td>\n",
       "      <td>..((((((((...................(((.....))).........</td>\n",
       "      <td>..((((((((...(((.............(((.....)))..(((....</td>\n",
       "      <td>..{{{((((({,,{{,,............(((.....)))|.((({...</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112.3(SLC26A2):c.483T&gt;C:p.Ser161Ser</td>\n",
       "      <td>coding_sequence</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.969680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.89147</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1836</td>\n",
       "      <td>solute carrier family 26 member 2</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>483</td>\n",
       "      <td>1738</td>\n",
       "      <td>0.218</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>CTC</td>\n",
       "      <td>GTG</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>CTG</td>\n",
       "      <td>)</td>\n",
       "      <td>(</td>\n",
       "      <td>...))))((</td>\n",
       "      <td>(..((((((</td>\n",
       "      <td>3</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TCC</td>\n",
       "      <td>ATC</td>\n",
       "      <td>GTG</td>\n",
       "      <td>ATCTCT</td>\n",
       "      <td>TCTGTG</td>\n",
       "      <td>ATCTCTGTG</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S4</td>\n",
       "      <td>S4</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_000112.3</td>\n",
       "      <td>5</td>\n",
       "      <td>149978075</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>269</td>\n",
       "      <td>2488</td>\n",
       "      <td>2220</td>\n",
       "      <td>8082</td>\n",
       "      <td>691</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>51</td>\n",
       "      <td>CAGTCCATTGCTTATTCCCTGCTGGCTGGCCAAGAACCTGTCTATG...</td>\n",
       "      <td>CAGTCCATTGCTTATTCCCTGCTGGCTGGCCAAGAACCTGTCTATG...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>1.8</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>1.490450</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1.45</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-14.7</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-21.88</td>\n",
       "      <td>13.43</td>\n",
       "      <td>20.31</td>\n",
       "      <td>80.37</td>\n",
       "      <td>0.065354</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>-14.79</td>\n",
       "      <td>-18.4</td>\n",
       "      <td>-21.72</td>\n",
       "      <td>15.23</td>\n",
       "      <td>21.76</td>\n",
       "      <td>76.45</td>\n",
       "      <td>0.085024</td>\n",
       "      <td>...................(((((((.((((((((.....))).))...</td>\n",
       "      <td>...................(((((((..((((.(.......)..))...</td>\n",
       "      <td>...................(((((((.((((((((.....))).))...</td>\n",
       "      <td>...................{((((((.((({((((.....|}}.})...</td>\n",
       "      <td>...................(((((((.((((((((.....))).))...</td>\n",
       "      <td>...................(((((((.......................</td>\n",
       "      <td>...................(((((((..(((((((.....))).))...</td>\n",
       "      <td>.,,................{((((((,(((,((((,..,,|,|,||...</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112.3(SLC26A2):c.423G&gt;A:p.Leu141Leu</td>\n",
       "      <td>coding_sequence</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.971716</td>\n",
       "      <td>2.0</td>\n",
       "      <td>246752.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0.914492</td>\n",
       "      <td>0.92406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1836</td>\n",
       "      <td>solute carrier family 26 member 2</td>\n",
       "      <td>SLC26A2</td>\n",
       "      <td>NM_000112</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>423</td>\n",
       "      <td>1798</td>\n",
       "      <td>0.191</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TAC</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>TGT</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>)).......</td>\n",
       "      <td>)).......</td>\n",
       "      <td>3</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CTA</td>\n",
       "      <td>GGT</td>\n",
       "      <td>TAC</td>\n",
       "      <td>GGTCTG</td>\n",
       "      <td>CTGTAC</td>\n",
       "      <td>GGTCTGTAC</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>L4</td>\n",
       "      <td>L4</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         NM_ID CHR        POS REF ALT Sense  CDS_Start  CDS_Stop  CDS_Length  \\\n",
       "0  NM_000112.3   5  149977727   T   A     +        269      2488        2220   \n",
       "1  NM_000112.3   5  149981141   T   C     +        269      2488        2220   \n",
       "2  NM_000112.3   5  149981291   G   A     +        269      2488        2220   \n",
       "3  NM_000112.3   5  149978135   T   C     +        269      2488        2220   \n",
       "4  NM_000112.3   5  149978075   G   A     +        269      2488        2220   \n",
       "\n",
       "   Trans_Length  Trans_POS REF_Base ALT_Base  WindowPosition  \\\n",
       "0          8082        343        T        A              51   \n",
       "1          8082       1816        T        C              51   \n",
       "2          8082       1966        G        A              51   \n",
       "3          8082        751        T        C              51   \n",
       "4          8082        691        G        A              51   \n",
       "\n",
       "                                        REF_Sequence  \\\n",
       "0  CATAACGTTTCACCCAGAGACTCAGCTGAAGGAAATGACAGTTATC...   \n",
       "1  GATCTTCCCAAAATGTGGAGTATTAGTAGAATGGATACAGTTATCT...   \n",
       "2  AAGAGTTCACTGCTTGGCTTGGTGGAAGAGTCTGAGGTCTTTGAAT...   \n",
       "3  TTTTTTGCCAGCATCATTTATTTTCTCTTGGGTACCTCCCGTCACA...   \n",
       "4  CAGTCCATTGCTTATTCCCTGCTGGCTGGCCAAGAACCTGTCTATG...   \n",
       "\n",
       "                                        ALT_Sequence  deltaMFE  deltaCFE  \\\n",
       "0  CATAACGTTTCACCCAGAGACTCAGCTGAAGGAAATGACAGTTATC...       0.9      5.70   \n",
       "1  GATCTTCCCAAAATGTGGAGTATTAGTAGAATGGATACAGTTATCT...      -2.2     -6.30   \n",
       "2  AAGAGTTCACTGCTTGGCTTGGTGGAAGAGTCTGAGGTCTTTGAAT...      -0.4      0.00   \n",
       "3  TTTTTTGCCAGCATCATTTATTTTCTCTTGGGTACCTCCCGTCACA...       0.2     12.48   \n",
       "4  CAGTCCATTGCTTATTCCCTGCTGGCTGGCCAAGAACCTGTCTATG...       0.0     -0.09   \n",
       "\n",
       "   deltaMEAFE  deltaEFE  MFEED  CFEED  MEAED      EFEED  deltaCD  deltaEND  \\\n",
       "0         3.9      0.60     40     36     24   9.908430     0.34      0.19   \n",
       "1        -3.3     -1.28     38     46     40  22.772300     1.68      2.89   \n",
       "2        -2.5     -0.20     14     12     16   0.395924    -0.84     -1.82   \n",
       "3         0.7      0.35     68     36      4   5.356520     2.78      3.67   \n",
       "4         1.8      0.16      0     20      4   1.490450     1.80      1.45   \n",
       "\n",
       "   REF_mfeValue  REF_cfeValue  REF_meafeValue  REF_efeValue  REF_cdValue  \\\n",
       "0         -18.1          -9.4           -10.2        -20.33        22.62   \n",
       "1         -17.7         -11.3           -16.6        -20.70        17.00   \n",
       "2         -26.9         -19.9           -24.2        -30.09        16.50   \n",
       "3         -20.2         -17.8           -18.2        -22.46        17.01   \n",
       "4         -20.2         -14.7           -20.2        -21.88        13.43   \n",
       "\n",
       "   REF_endValue  REF_meaValue  REF_freqMfeEnsemble  ALT_mfeValue  \\\n",
       "0         29.89         63.62             0.026939         -17.2   \n",
       "1         25.15         75.20             0.007678         -19.9   \n",
       "2         25.12         74.62             0.005638         -27.3   \n",
       "3         24.82         74.72             0.025643         -20.0   \n",
       "4         20.31         80.37             0.065354         -20.2   \n",
       "\n",
       "   ALT_cfeValue  ALT_meafeValue  ALT_efeValue  ALT_cdValue  ALT_endValue  \\\n",
       "0         -3.70            -6.3        -19.73        22.96         30.08   \n",
       "1        -17.60           -19.9        -21.98        18.68         28.04   \n",
       "2        -19.90           -26.7        -30.29        15.66         23.30   \n",
       "3         -5.32           -17.5        -22.11        19.79         28.49   \n",
       "4        -14.79           -18.4        -21.72        15.23         21.76   \n",
       "\n",
       "   ALT_meaValue  ALT_freqMfeEnsemble  \\\n",
       "0         62.60             0.016616   \n",
       "1         72.11             0.034310   \n",
       "2         76.75             0.007827   \n",
       "3         68.23             0.032343   \n",
       "4         76.45             0.085024   \n",
       "\n",
       "                                    REF_mfeStructure  \\\n",
       "0  ......((((((((((((.....(((((..........)))))......   \n",
       "1  .......(((.....)))(((((((((((..((((.(((((....(...   \n",
       "2  .((((((...(((((((((((........(((((((((.((((((....   \n",
       "3  ..((((((((..(((((............(((.....)))..((((...   \n",
       "4  ...................(((((((.((((((((.....))).))...   \n",
       "\n",
       "                                    REF_cfeStructure  \\\n",
       "0  ......(((((.....)))))...((.((((..................   \n",
       "1  ..................(((((((((((...(((.(((((........   \n",
       "2  ..(((((...(((((((((((........(((((.(((...........   \n",
       "3  ..((((((((...................(((.....)))..((((...   \n",
       "4  ...................(((((((..((((.(.......)..))...   \n",
       "\n",
       "                                  REF_meafeStructure  \\\n",
       "0  ......(((((.....)))))...((.((((.............((...   \n",
       "1  .......(((.....)))(((((((((((...(((.(((((....(...   \n",
       "2  ..(((((...(((((((((((........(((((((((((.((......   \n",
       "3  ..((((((((...(((.............(((.....)))..((((...   \n",
       "4  ...................(((((((.((((((((.....))).))...   \n",
       "\n",
       "                                    REF_efeStructure  \\\n",
       "0  ......(((((.,,,,||}}}..,(((({{(....,..|}}},.{{...   \n",
       "1  ......,(({....,}))(((((((((((..,(((.(((((....{...   \n",
       "2  .,(((({...(((((((((((,......,(((((((((,({,,......   \n",
       "3  ..{{{(((((,.,{{{,...........,(((.....)))}.((((...   \n",
       "4  ...................{((((((.((({((((.....|}}.})...   \n",
       "\n",
       "                                    ALT_mfeStructure  \\\n",
       "0  .............(((((.......((((.((((((....))).))...   \n",
       "1  .............(((((((.((..(((((((((((.(((....))...   \n",
       "2  .((((((...(((((((((((........(((((((((((.........   \n",
       "3  ........((((.................(((.....)))(((.((...   \n",
       "4  ...................(((((((.((((((((.....))).))...   \n",
       "\n",
       "                                    ALT_cfeStructure  \\\n",
       "0  ......(((((.....)))))............................   \n",
       "1  .............(((((((.....(((((((((((.(((....))...   \n",
       "2  ..(((((...(((((((((((........(((((.(((...........   \n",
       "3  ..((((((((...................(((.....))).........   \n",
       "4  ...................(((((((.......................   \n",
       "\n",
       "                                  ALT_meafeStructure  \\\n",
       "0  ......(((((.....)))))...((.((((((...........))...   \n",
       "1  .............(((((((.((..(((((((((((.(((....))...   \n",
       "2  ..(((((...(((((((((((........(((((((((((.........   \n",
       "3  ..((((((((...(((.............(((.....)))..(((....   \n",
       "4  ...................(((((((..(((((((.....))).))...   \n",
       "\n",
       "                                    ALT_efeStructure     GENE  \\\n",
       "0  ......((({{..,,,||}}}...{(((({({,{,{....},}.||...  SLC26A2   \n",
       "1  ........,....{((((((.({,,(((((((((({.(((....))...  SLC26A2   \n",
       "2  .,(((({...(((((((((((.......,(((((((((,{,,,......  SLC26A2   \n",
       "3  ..{{{((((({,,{{,,............(((.....)))|.((({...  SLC26A2   \n",
       "4  .,,................{((((((,(((,((((,..,,|,|,||...  SLC26A2   \n",
       "\n",
       "                                         HGVS      LOC_IN_GENE  \\\n",
       "0     NM_000112.3(SLC26A2):c.75T>A:p.Ser25Ser  coding_sequence   \n",
       "1  NM_000112.3(SLC26A2):c.1548T>C:p.Phe516Phe  coding_sequence   \n",
       "2  NM_000112.3(SLC26A2):c.1698G>A:p.Val566Val  coding_sequence   \n",
       "3   NM_000112.3(SLC26A2):c.483T>C:p.Ser161Ser  coding_sequence   \n",
       "4   NM_000112.3(SLC26A2):c.423G>A:p.Leu141Leu  coding_sequence   \n",
       "\n",
       "           EFFECT  DIST_TO_CODING_REGION DIST_TO_SPLICE_SITE IMPACT  \\\n",
       "0  synonymous SNV                    NaN                 NaN    LOW   \n",
       "1  synonymous SNV                    NaN                 NaN    LOW   \n",
       "2  synonymous SNV                    NaN                 NaN    LOW   \n",
       "3  synonymous SNV                    NaN                 NaN    LOW   \n",
       "4  synonymous SNV                    NaN                 NaN    LOW   \n",
       "\n",
       "   gnomAD3_WG_AC  gnomAD3_WG_AN  gnomAD3_WG_AF  gnomAD3_WG_nhomalt  \\\n",
       "0            NaN            NaN            NaN                 NaN   \n",
       "1            NaN            NaN            NaN                 NaN   \n",
       "2            NaN            NaN            NaN                 NaN   \n",
       "3            NaN            NaN            NaN                 NaN   \n",
       "4            NaN            NaN            NaN                 NaN   \n",
       "\n",
       "   gnomAD3_Coverage gnomAD3_Quality  gnomAD3_WG_InbreedingCoeff  \\\n",
       "0              True            PASS                         NaN   \n",
       "1              True            PASS                         NaN   \n",
       "2              True            PASS                         NaN   \n",
       "3              True            PASS                         NaN   \n",
       "4              True            PASS                         NaN   \n",
       "\n",
       "   gnomAD3_CVGE_OVER_20  gnomAD2_EX_AC  gnomAD2_EX_AN  gnomAD2_EX_AF  \\\n",
       "0              0.973474            NaN            NaN            NaN   \n",
       "1              0.965803            NaN            NaN            NaN   \n",
       "2              0.954799            NaN            NaN            NaN   \n",
       "3              0.969680            NaN            NaN            NaN   \n",
       "4              0.971716            2.0       246752.0       0.000008   \n",
       "\n",
       "   gnomAD2_EX_nhomalt  gnomAD2_Coverage gnomAD2_Quality  \\\n",
       "0                 NaN              True            PASS   \n",
       "1                 NaN              True            PASS   \n",
       "2                 NaN              True            PASS   \n",
       "3                 NaN              True            PASS   \n",
       "4                 0.0              True            PASS   \n",
       "\n",
       "   gnomAD2_EX_InbreedingCoeff  gnomAD2_EX_rf_tp_probability  \\\n",
       "0                         NaN                           NaN   \n",
       "1                         NaN                           NaN   \n",
       "2                         NaN                           NaN   \n",
       "3                         NaN                           NaN   \n",
       "4                      0.0061                      0.914492   \n",
       "\n",
       "   gnomAD2_CVGE_OVER_20 gnomAD3_WG_vep  entrezgene  \\\n",
       "0               0.99905            NaN        1836   \n",
       "1               0.99800            NaN        1836   \n",
       "2               0.99916            NaN        1836   \n",
       "3               0.89147            NaN        1836   \n",
       "4               0.92406            NaN        1836   \n",
       "\n",
       "                                name   symbol NM_ID_noVersion  is_MANE  \\\n",
       "0  solute carrier family 26 member 2  SLC26A2       NM_000112     True   \n",
       "1  solute carrier family 26 member 2  SLC26A2       NM_000112     True   \n",
       "2  solute carrier family 26 member 2  SLC26A2       NM_000112     True   \n",
       "3  solute carrier family 26 member 2  SLC26A2       NM_000112     True   \n",
       "4  solute carrier family 26 member 2  SLC26A2       NM_000112     True   \n",
       "\n",
       "   is_LongestCDS  is_LongestTrans  snpeffTYPE  PosRelative2Start  \\\n",
       "0           True             True  synonymous                 75   \n",
       "1           True             True  synonymous               1548   \n",
       "2           True             True  synonymous               1698   \n",
       "3           True             True  synonymous                483   \n",
       "4           True             True  synonymous                423   \n",
       "\n",
       "   PosRelative2Stop  PosRelative2StartPct PrecedingBase TrailingBase  \\\n",
       "0              2146                 0.034             C            G   \n",
       "1               673                 0.697             T            G   \n",
       "2               523                 0.765             T            T   \n",
       "3              1738                 0.218             C            G   \n",
       "4              1798                 0.191             T            T   \n",
       "\n",
       "  PrecedingBases TrailingBases SNVContext TribaseContext REF_BondMFE  \\\n",
       "0            ATC           GGG        T>A            CTG           )   \n",
       "1            GTT           GTT        T>C            TTG           .   \n",
       "2            TGT           TCT        G>A            TGT           .   \n",
       "3            CTC           GTG        T>C            CTG           )   \n",
       "4            TCT           TAC        G>A            TGT           .   \n",
       "\n",
       "  ALT_BondMFE REF_LocalMFEStructure ALT_LocalMFEStructure  CodonPosition  \\\n",
       "0           )             .))))))..             .))))....              3   \n",
       "1           )             ((.....))             ))))))).)              3   \n",
       "2           (             .......).             ((((.....              3   \n",
       "3           (             ...))))((             (..((((((              3   \n",
       "4           .             )).......             )).......              3   \n",
       "\n",
       "  REF_Codon ALT_Codon PrecedingCodon TrailingCodon PrecedingBicodon  \\\n",
       "0       TCT       TCA            CCA           GGG           CCATCT   \n",
       "1       TTT       TTC            TGG           GTT           TGGTTT   \n",
       "2       GTG       GTA            TCT           TCT           TCTGTG   \n",
       "3       TCT       TCC            ATC           GTG           ATCTCT   \n",
       "4       CTG       CTA            GGT           TAC           GGTCTG   \n",
       "\n",
       "  TrailingBicodon TricodonContext REF_AminoAcid ALT_AminoAcid  \\\n",
       "0          TCTGGG       CCATCTGGG             S             S   \n",
       "1          TTTGTT       TGGTTTGTT             F             F   \n",
       "2          GTGTCT       TCTGTGTCT             V             V   \n",
       "3          TCTGTG       ATCTCTGTG             S             S   \n",
       "4          CTGTAC       GGTCTGTAC             L             L   \n",
       "\n",
       "  REF_AminoAcid_sub ALT_AminoAcid_sub  manualTYPE  DupAlignment_select  \\\n",
       "0                S4                S4  synonymous                False   \n",
       "1                 F                 F  synonymous                False   \n",
       "2                 V                 V  synonymous                False   \n",
       "3                S4                S4  synonymous                False   \n",
       "4                L4                L4  synonymous                False   \n",
       "\n",
       "   maxGeneCount  \n",
       "0             1  \n",
       "1             1  \n",
       "2             1  \n",
       "3             1  \n",
       "4             1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_variant_syn_df = pd.read_csv(rnas_variant_syn_filename,\n",
    "                                  sep=\"\\t\",\n",
    "                                  dtype={\"CHR\":str})\n",
    "rnas_variant_syn_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2e4901ab-80af-4f41-b212-45c6d33e3d8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_ID\n",
      "CHR\n",
      "POS\n",
      "REF\n",
      "ALT\n",
      "Sense\n",
      "CDS_Start\n",
      "CDS_Stop\n",
      "CDS_Length\n",
      "Trans_Length\n",
      "Trans_POS\n",
      "REF_Base\n",
      "ALT_Base\n",
      "WindowPosition\n",
      "REF_Sequence\n",
      "ALT_Sequence\n",
      "deltaMFE\n",
      "deltaCFE\n",
      "deltaMEAFE\n",
      "deltaEFE\n",
      "MFEED\n",
      "CFEED\n",
      "MEAED\n",
      "EFEED\n",
      "deltaCD\n",
      "deltaEND\n",
      "REF_mfeValue\n",
      "REF_cfeValue\n",
      "REF_meafeValue\n",
      "REF_efeValue\n",
      "REF_cdValue\n",
      "REF_endValue\n",
      "REF_meaValue\n",
      "REF_freqMfeEnsemble\n",
      "ALT_mfeValue\n",
      "ALT_cfeValue\n",
      "ALT_meafeValue\n",
      "ALT_efeValue\n",
      "ALT_cdValue\n",
      "ALT_endValue\n",
      "ALT_meaValue\n",
      "ALT_freqMfeEnsemble\n",
      "REF_mfeStructure\n",
      "REF_cfeStructure\n",
      "REF_meafeStructure\n",
      "REF_efeStructure\n",
      "ALT_mfeStructure\n",
      "ALT_cfeStructure\n",
      "ALT_meafeStructure\n",
      "ALT_efeStructure\n",
      "GENE\n",
      "HGVS\n",
      "LOC_IN_GENE\n",
      "EFFECT\n",
      "DIST_TO_CODING_REGION\n",
      "DIST_TO_SPLICE_SITE\n",
      "IMPACT\n",
      "gnomAD3_WG_AC\n",
      "gnomAD3_WG_AN\n",
      "gnomAD3_WG_AF\n",
      "gnomAD3_WG_nhomalt\n",
      "gnomAD3_Coverage\n",
      "gnomAD3_Quality\n",
      "gnomAD3_WG_InbreedingCoeff\n",
      "gnomAD3_CVGE_OVER_20\n",
      "gnomAD2_EX_AC\n",
      "gnomAD2_EX_AN\n",
      "gnomAD2_EX_AF\n",
      "gnomAD2_EX_nhomalt\n",
      "gnomAD2_Coverage\n",
      "gnomAD2_Quality\n",
      "gnomAD2_EX_InbreedingCoeff\n",
      "gnomAD2_EX_rf_tp_probability\n",
      "gnomAD2_CVGE_OVER_20\n",
      "gnomAD3_WG_vep\n",
      "entrezgene\n",
      "name\n",
      "symbol\n",
      "NM_ID_noVersion\n",
      "is_MANE\n",
      "is_LongestCDS\n",
      "is_LongestTrans\n",
      "snpeffTYPE\n",
      "PosRelative2Start\n",
      "PosRelative2Stop\n",
      "PosRelative2StartPct\n",
      "PrecedingBase\n",
      "TrailingBase\n",
      "PrecedingBases\n",
      "TrailingBases\n",
      "SNVContext\n",
      "TribaseContext\n",
      "REF_BondMFE\n",
      "ALT_BondMFE\n",
      "REF_LocalMFEStructure\n",
      "ALT_LocalMFEStructure\n",
      "CodonPosition\n",
      "REF_Codon\n",
      "ALT_Codon\n",
      "PrecedingCodon\n",
      "TrailingCodon\n",
      "PrecedingBicodon\n",
      "TrailingBicodon\n",
      "TricodonContext\n",
      "REF_AminoAcid\n",
      "ALT_AminoAcid\n",
      "REF_AminoAcid_sub\n",
      "ALT_AminoAcid_sub\n",
      "manualTYPE\n",
      "DupAlignment_select\n",
      "maxGeneCount\n"
     ]
    }
   ],
   "source": [
    "for col in rnas_variant_syn_df.columns:\n",
    "    print(col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b049501a-3ae2-4980-b513-1329a6a4df3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22050128, 111)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_variant_syn_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaa2d71f-4b9f-404f-9a3b-6f0925b95e43",
   "metadata": {},
   "source": [
    "## Remove near-end variants and single degeneracy codons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86e0c16b-b99f-4fc8-995d-3107f8c490b1",
   "metadata": {},
   "source": [
    "Select sites removed a bit from the start codon:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f13dbfa-20b8-411f-86b4-9c3ba649e0e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_variant_syn_noedge_df = rnas_variant_syn_df.query(\"PosRelative2Start > 51 & \" + \n",
    "                                                       \"PosRelative2Stop > 51\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "004e6870-650b-451c-af1d-93a92a746a54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20789723, 111)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_variant_syn_noedge_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cddf425-c32a-4f80-ae2e-c7f0830ab339",
   "metadata": {},
   "source": [
    "Filter on REF and ALT amino acids for those with synonyms (should already be selected) and no STOP codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9fbbd9b1-5592-47c1-8028-373e58b40f96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20789723, 111)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_variant_syn_noedge_nostop_df = rnas_variant_syn_noedge_df.query((\"REF_AminoAcid in @ccv.aa_with_syn_noSTOP & \" + \n",
    "                                                                      \"ALT_AminoAcid in @ccv.aa_with_syn_noSTOP\"))\n",
    "rnas_variant_syn_noedge_nostop_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "43b6c0a8-5183-4d39-86ac-5864977dfc14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51    20789723\n",
       "Name: WindowPosition, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_variant_syn_noedge_nostop_df[\"WindowPosition\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35566ebf-a328-4732-907c-7c535d041e63",
   "metadata": {},
   "source": [
    "## Remove records with STOP codons appearing within sequence context"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c8dc99bb-f289-4629-9bb1-c875611d36e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_codon_context_noSTOP = rnas_variant_syn_noedge_nostop_df.apply(filter_contexts_for_codons, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9122c540-abee-4315-8168-8a88e8cb68da",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_variant_syn_noedge_noSTOP_df= rnas_variant_syn_noedge_nostop_df[variant_codon_context_noSTOP]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "34b9a9a5-97e8-4149-8c41-19e14e542ced",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20787246, 111)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_variant_syn_noedge_noSTOP_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16ce5103-8678-49ea-9d73-601da87d0f43",
   "metadata": {},
   "source": [
    "## Subset records within the same REF/ALT Amino Acid sub class "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c581540-4d94-43e1-ba20-3303cf3de123",
   "metadata": {},
   "source": [
    "Check how many sSNVs switch between different AA_sub classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dc1d7f90-8347-4118-9fab-c53c383c1018",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1091987"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(rnas_variant_syn_noedge_noSTOP_df[\"REF_AminoAcid_sub\"] != rnas_variant_syn_noedge_noSTOP_df[\"ALT_AminoAcid_sub\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5e884821-10c6-4aa0-b4e5-414f77803a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnas_variant_syn_noedge_noSTOP_CP3_df = rnas_variant_syn_noedge_noSTOP_df.query(\"REF_AminoAcid_sub == ALT_AminoAcid_sub\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4ebfeed0-4f25-47bc-a97a-e4780af28b09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19695259, 111)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnas_variant_syn_noedge_noSTOP_CP3_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7df9b108-b12c-4ecd-9a51-37369569fd79",
   "metadata": {},
   "source": [
    "## Save filtered files\n",
    "\n",
    "Save per Amino Acid files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "20b75b22-37db-4ffd-b058-347224dfe955",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NM_ID', 'CDS_Start', 'CDS_Stop', 'CDS_Length', 'Trans_Length', 'Trans_POS', 'REF_Base', 'ALT_Base', 'ALT_Sequence', 'deltaMFE', 'deltaCFE', 'deltaMEAFE', 'deltaEFE', 'MFEED', 'CFEED', 'MEAED', 'EFEED', 'deltaCD', 'deltaEND', 'REF_mfeValue', 'REF_cfeValue', 'REF_meafeValue', 'REF_efeValue', 'REF_cdValue', 'REF_endValue', 'REF_meaValue', 'REF_freqMfeEnsemble', 'ALT_mfeValue', 'ALT_cfeValue', 'ALT_meafeValue', 'ALT_efeValue', 'ALT_cdValue', 'ALT_endValue', 'ALT_meaValue', 'ALT_freqMfeEnsemble', 'REF_mfeStructure', 'REF_cfeStructure', 'REF_meafeStructure', 'REF_efeStructure', 'ALT_mfeStructure', 'ALT_cfeStructure', 'ALT_meafeStructure', 'ALT_efeStructure', 'GENE', 'HGVS', 'LOC_IN_GENE', 'EFFECT', 'DIST_TO_CODING_REGION', 'DIST_TO_SPLICE_SITE', 'IMPACT', 'gnomAD3_WG_AC', 'gnomAD3_WG_AN', 'gnomAD3_WG_AF', 'gnomAD3_WG_nhomalt', 'gnomAD3_Coverage', 'gnomAD3_Quality', 'gnomAD3_WG_InbreedingCoeff', 'gnomAD3_CVGE_OVER_20', 'gnomAD2_EX_AC', 'gnomAD2_EX_AN', 'gnomAD2_EX_AF', 'gnomAD2_EX_nhomalt', 'gnomAD2_Coverage', 'gnomAD2_Quality', 'gnomAD2_EX_InbreedingCoeff', 'gnomAD2_EX_rf_tp_probability', 'gnomAD2_CVGE_OVER_20', 'gnomAD3_WG_vep', 'entrezgene', 'name', 'symbol', 'NM_ID_noVersion', 'is_MANE', 'is_LongestCDS', 'is_LongestTrans', 'snpeffTYPE', 'PosRelative2Start', 'PosRelative2Stop', 'PosRelative2StartPct', 'REF_BondMFE', 'ALT_BondMFE', 'REF_LocalMFEStructure', 'ALT_LocalMFEStructure', 'manualTYPE', 'DupAlignment_select', 'maxGeneCount']\n"
     ]
    }
   ],
   "source": [
    "key_columns = [\"CHR\", \"POS\", \"REF\", \"ALT\"]\n",
    "seq_columns = [\"Sense\",\n",
    "               \"REF_AminoAcid\", \"ALT_AminoAcid\", \n",
    "               \"REF_AminoAcid_sub\", \"ALT_AminoAcid_sub\",\n",
    "               \"REF_Codon\", \"ALT_Codon\",\n",
    "               \"CodonPosition\", \"WindowPosition\",\n",
    "               \"PrecedingBase\", \"TrailingBase\",\n",
    "               \"PrecedingBases\", \"TrailingBases\",\n",
    "               \"SNVContext\", \"TribaseContext\",\n",
    "               \"PrecedingCodon\", \"TrailingCodon\", \n",
    "               \"PrecedingBicodon\", \"TrailingBicodon\",\n",
    "               \"TricodonContext\", \n",
    "               \"REF_Sequence\"]\n",
    "other_columns = [x for x in rnas_variant_syn_noedge_noSTOP_CP3_df.columns\n",
    "                 if x not in key_columns+seq_columns]\n",
    "print(other_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "440f63a1-2abd-4227-8b3f-584ee3e14f5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\n",
      "(2097117, 111)\n",
      "C\n",
      "(230056, 111)\n",
      "D\n",
      "(483497, 111)\n",
      "E\n",
      "(723696, 111)\n",
      "F\n",
      "(369312, 111)\n",
      "G\n",
      "(1979203, 111)\n",
      "H\n",
      "(266642, 111)\n",
      "I\n",
      "(885905, 111)\n",
      "K\n",
      "(580497, 111)\n",
      "L2\n",
      "(211848, 111)\n",
      "L4\n",
      "(2362456, 111)\n",
      "N\n",
      "(366106, 111)\n",
      "P\n",
      "(1894828, 111)\n",
      "Q\n",
      "(481529, 111)\n",
      "R2\n",
      "(238434, 111)\n",
      "R4\n",
      "(976715, 111)\n",
      "S2\n",
      "(328278, 111)\n",
      "S4\n",
      "(1521162, 111)\n",
      "T\n",
      "(1613996, 111)\n",
      "V\n",
      "(1812349, 111)\n",
      "Y\n",
      "(271633, 111)\n"
     ]
    }
   ],
   "source": [
    "for amin, amin_variant_filtered_df in rnas_variant_syn_noedge_noSTOP_CP3_df.groupby(\"REF_AminoAcid_sub\") :\n",
    "    print(amin)\n",
    "    print(amin_variant_filtered_df.shape)\n",
    "    \n",
    "    #-extract just key columns and sequence information\n",
    "    amin_variant_filtered_seq_df = amin_variant_filtered_df[key_columns+seq_columns]\n",
    "    #-extract key columns and non-seq columns\n",
    "    amin_variant_filtered_nonseq_df = amin_variant_filtered_df[key_columns+other_columns]\n",
    "    \n",
    "    #Write out\n",
    "    amin_variant_filtered_seq_df.to_csv(rnas_variant_syn_out[amin][\"seq_filename\"],\n",
    "                                        sep=\"\\t\",\n",
    "                                        index=False)\n",
    "    amin_variant_filtered_nonseq_df.to_csv(rnas_variant_syn_out[amin][\"nonseq_filename\"],\n",
    "                                           sep=\"\\t\",\n",
    "                                           index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:codon_context]",
   "language": "python",
   "name": "conda-env-codon_context-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
