{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Purpose**\n",
    "\n",
    "* Determine the cohorts and mutational signatures where there is an enrichment of low $\\Delta$ context scores over expected\n",
    "* Determine the gene sets for each cohort that have significant enrichment among genes appearing in high context effect variant sets"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files - read"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enrichment of high context effect variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresholds_5_brca_ucec = \"../../data/4_tcga_analysis/scores_enriched_on_5quantile_cohortBRCA_UCEC_TCGA_AC0bg_ConNegCP3.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GSEA outputs for each cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-observed\n",
    "gsea_brca_5q_filename = \"../../data/4_tcga_analysis/gseapy_ora_kegg_go_cohortBRCA_AC0_5q_CG_GA.tsv\"\n",
    "gsea_ucec_5q_filename = \"../../data/4_tcga_analysis/gseapy_ora_kegg_go_cohortUCEC_AC0_5q_CT_GA.tsv\"\n",
    "gsea_brca_ucec_5q_filename = \"../../data/4_tcga_analysis/gseapy_ora_kegg_go_cohortBRCA_UCEC_AC0_5q_CG_GA_CT.tsv\"\n",
    "#-sampling\n",
    "gsea_brca_5q_sampled_filename = \"../../data/4_tcga_analysis/gseapy_sampled_ora_kegg_go_cohortBRCA_n100_AC0_5q_CG_GA.tsv\"\n",
    "gsea_ucec_5q_sampled_filename = \"../../data/4_tcga_analysis/gseapy_sampled_ora_kegg_go_cohortUCEC_n100_AC0_5q_CT_GA.tsv\"\n",
    "gsea_brca_ucec_5q_sampled_filename = \"../../data/4_tcga_analysis/gseapy_sampled_ora_kegg_go_cohortBRCA_UCEC_n100_AC0_5q_CG_GA_CT.tsv\"\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files - write"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_summary_filename = \"../../data/4_tcga_analysis/gseapy_ora_summary_kegg_go_cohortBRCA_UCEC_n100_AC0_5q_CT_CG_GA.tsv\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enriched scores in TCGA cohorts\n",
    "\n",
    "Loop through threshold files and print enriched results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Quantile: 5\n",
      "          cohort       group context  threshold_quantile  threshold_value  \\\n",
      "3           BRCA  SNVContext     C>G                0.05        -0.699761   \n",
      "7           BRCA  SNVContext     G>A                0.05        -1.709938   \n",
      "13          UCEC  SNVContext     C>T                0.05        -1.572423   \n",
      "16          UCEC  SNVContext     G>A                0.05        -1.709938   \n",
      "21  grouped_TCGA  SNVContext     C>G                0.05        -0.699761   \n",
      "22  grouped_TCGA  SNVContext     C>T                0.05        -1.572423   \n",
      "25  grouped_TCGA  SNVContext     G>A                0.05        -1.709938   \n",
      "\n",
      "   comparison  cohort_number_in  cohort_number_out  cohort_expected_in  \\\n",
      "3       below               134                911                  53   \n",
      "7       below               197               2556                 138   \n",
      "13      below               958              11759                 652   \n",
      "16      below               643              10396                 555   \n",
      "21      below               156               1291                  74   \n",
      "22      below              1105              14555                 803   \n",
      "25      below               837              12920                 691   \n",
      "\n",
      "    cohort_expected_out   test_stat       test_pv  \n",
      "3                   992  130.406364  3.339179e-30  \n",
      "7                  2615   26.555804  2.560344e-07  \n",
      "13                12065  151.374458  8.680499e-35  \n",
      "16                10484   14.691803  1.265958e-04  \n",
      "21                 1373   95.762170  1.295477e-22  \n",
      "22                14857  119.717868  7.292912e-28  \n",
      "25                13066   32.479456  1.204592e-08  \n"
     ]
    }
   ],
   "source": [
    "quantiles = [5]\n",
    "qfiles = [thresholds_5_brca_ucec]\n",
    "pvalue_cutoff = 0.05/(9*3)\n",
    "\n",
    "#test stat filter\n",
    "direction_filter = \"cohort_number_in > cohort_expected_in\"\n",
    "\n",
    "for i,q in enumerate(quantiles) :\n",
    "    tcga_quantile_df = pd.read_csv(qfiles[i],\n",
    "                                   sep=\"\\t\")\n",
    "    print(\"Quantile:\",q)\n",
    "    print(tcga_quantile_df.query(\"test_pv < @pvalue_cutoff &\"+\n",
    "                                 direction_filter))\n",
    "    "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pathway enrichment\n",
    "\n",
    "* Load the ORA results per cohort, per threshold\n",
    "* Load corresponding random-sampled result (filter to p-value threshold)\n",
    "* Mark ones that overlap\n",
    "* Combine into array and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pv_threshold = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_file_dict = {\"BRCA_5q\": gsea_brca_5q_filename,\n",
    "                  \"UCEC_5q\": gsea_ucec_5q_filename,\n",
    "                  \"BRCA_UCEC_5q\": gsea_brca_ucec_5q_filename}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_sampled_file_dict = {\"BRCA_5q\": gsea_brca_5q_sampled_filename,\n",
    "                          \"UCEC_5q\": gsea_ucec_5q_sampled_filename,\n",
    "                          \"BRCA_UCEC_5q\": gsea_brca_ucec_5q_sampled_filename}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA_5q\n",
      "(2, 10)\n",
      "(9, 8)\n",
      "(2, 10)\n",
      "                                                  Term Overlap  \\\n",
      "224  Somatic Recombination Of Immunoglobulin Genes ...     4/9   \n",
      "225                     Isotype Switching (GO:0045190)    4/11   \n",
      "\n",
      "     Adjusted P-value  Odds Ratio  \n",
      "224          0.014072   49.491824  \n",
      "225          0.017963   35.347709  \n",
      "UCEC_5q\n",
      "(2, 10)\n",
      "(191, 8)\n",
      "(2, 10)\n",
      "                                 Term Overlap  Adjusted P-value  Odds Ratio\n",
      "0    Herpes simplex virus 1 infection  75/498      2.780892e-09    2.627076\n",
      "296  DNA Damage Response (GO:0006974)  49/384      2.022286e-02    2.132688\n",
      "BRCA_UCEC_5q\n",
      "(14, 10)\n",
      "(345, 8)\n",
      "(13, 10)\n",
      "                                                  Term Overlap  \\\n",
      "0                     Herpes simplex virus 1 infection  89/498   \n",
      "1                             Homologous recombination   12/41   \n",
      "301             Nuclear Pore Organization (GO:0006999)    8/14   \n",
      "302                   DNA Damage Response (GO:0006974)  57/384   \n",
      "303     Protein Alpha-1,2-Demannosylation (GO:0036508)    8/16   \n",
      "304           Nucleocytoplasmic Transport (GO:0006913)   14/50   \n",
      "305                     Nuclear Transport (GO:0051169)   12/39   \n",
      "306         Nuclear Pore Complex Assembly (GO:0051292)    6/10   \n",
      "307  Cellular Response To Interferon-Alpha (GO:0035...     5/7   \n",
      "3                                   Lysine degradation   14/63   \n",
      "308                            DNA Repair (GO:0006281)  44/291   \n",
      "309          Peptidyl-Serine Modification (GO:0018209)  29/166   \n",
      "310       Peptidyl-Serine Phosphorylation (GO:0018105)  28/158   \n",
      "\n",
      "     Adjusted P-value  Odds Ratio  \n",
      "0        1.602496e-09    2.456442  \n",
      "1        1.375842e-02    4.549486  \n",
      "301      1.810183e-02   14.642686  \n",
      "302      2.218956e-02    1.937672  \n",
      "303      2.218956e-02   10.980815  \n",
      "304      3.567856e-02    4.279182  \n",
      "305      3.567856e-02    4.887019  \n",
      "306      3.567856e-02   16.455090  \n",
      "307      4.260927e-02   27.411730  \n",
      "3        4.472505e-02    3.141654  \n",
      "308      4.663958e-02    1.973158  \n",
      "309      4.731639e-02    2.337464  \n",
      "310      4.731639e-02    2.377857  \n"
     ]
    }
   ],
   "source": [
    "#Store dataframes\n",
    "gsea_dfs = []\n",
    "\n",
    "for label, gs_filename in gsea_file_dict.items() :\n",
    "    \n",
    "    print(label)\n",
    "    gsea_df = pd.read_csv(gs_filename,\n",
    "                          sep=\"\\t\")\n",
    "    gsea_sampled_df = pd.read_csv(gsea_sampled_file_dict[label],\n",
    "                                   sep=\"\\t\")\n",
    "    \n",
    "    gsea_filter_df = gsea_df.query(\"`Adjusted P-value` < @adj_pv_threshold\")\n",
    "    print(gsea_filter_df.shape)\n",
    "    \n",
    "    #Grab the matching sampled table\n",
    "    gsea_sampled_filter = gsea_sampled_df.query(\"`Adjusted P-value` < @adj_pv_threshold\")\n",
    "    print(gsea_sampled_filter.shape)\n",
    "    sampling_count = gsea_sampled_filter.value_counts(\"Term\")\n",
    "    sampling_terms = sampling_count.index.to_list()\n",
    "    \n",
    "    #Look at table with matching terms removed\n",
    "    gsea_notSampled = gsea_filter_df[~gsea_filter_df[\"Term\"].isin(sampling_terms)]\n",
    "    print(gsea_notSampled.shape)\n",
    "    print(gsea_notSampled.sort_values(\"Adjusted P-value\")[[\"Term\",\"Overlap\",\n",
    "                                                           \"Adjusted P-value\",\"Odds Ratio\"]])\n",
    "    \n",
    "    #Mark overlap with random sampling\n",
    "    #-note: we only retained results when p<0.05, so it doesn't mark\n",
    "    # every pathway returned in the random sampling, at every p-value\n",
    "    gsea_df[\"cohort\"] = label\n",
    "    gsea_df[\"number_sampled_in_random\"] = 0\n",
    "    for i,g in gsea_filter_df.iterrows() :\n",
    "        if g[\"Term\"] in sampling_terms :\n",
    "            gsea_df.at[i,\"number_sampled_in_random\"] = sampling_count[g[\"Term\"]]\n",
    "        else :\n",
    "            pass\n",
    "    gsea_dfs.append(gsea_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Combine data frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "      <th>cohort</th>\n",
       "      <th>number_sampled_in_random</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>8/171</td>\n",
       "      <td>0.006622</td>\n",
       "      <td>0.958501</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.050291</td>\n",
       "      <td>15.304373</td>\n",
       "      <td>SEC61A1;SEC24A;MAN1A2;UBE4B;DNAJC5G;STT3B;UGGT...</td>\n",
       "      <td>BRCA_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>Non-homologous end-joining</td>\n",
       "      <td>2/13</td>\n",
       "      <td>0.017927</td>\n",
       "      <td>0.958501</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11.174432</td>\n",
       "      <td>44.937142</td>\n",
       "      <td>RAD50;LIG4</td>\n",
       "      <td>BRCA_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>Ubiquitin mediated proteolysis</td>\n",
       "      <td>6/140</td>\n",
       "      <td>0.025962</td>\n",
       "      <td>0.958501</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.769318</td>\n",
       "      <td>10.111073</td>\n",
       "      <td>HERC2;UBR5;UBE4B;BRCA1;BTRC;CBL</td>\n",
       "      <td>BRCA_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>Homologous recombination</td>\n",
       "      <td>3/41</td>\n",
       "      <td>0.028089</td>\n",
       "      <td>0.958501</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.860584</td>\n",
       "      <td>17.363799</td>\n",
       "      <td>RAD50;RAD51C;BRCA1</td>\n",
       "      <td>BRCA_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>PPAR signaling pathway</td>\n",
       "      <td>4/74</td>\n",
       "      <td>0.031491</td>\n",
       "      <td>0.958501</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.523450</td>\n",
       "      <td>12.184308</td>\n",
       "      <td>HMGCS1;ACOX1;EHHADH;PLIN2</td>\n",
       "      <td>BRCA_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene_set                                         Term Overlap  \\\n",
       "0  KEGG_2021_Human  Protein processing in endoplasmic reticulum   8/171   \n",
       "1  KEGG_2021_Human                   Non-homologous end-joining    2/13   \n",
       "2  KEGG_2021_Human               Ubiquitin mediated proteolysis   6/140   \n",
       "3  KEGG_2021_Human                     Homologous recombination    3/41   \n",
       "4  KEGG_2021_Human                       PPAR signaling pathway    4/74   \n",
       "\n",
       "    P-value  Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
       "0  0.006622          0.958501            0                     0    3.050291   \n",
       "1  0.017927          0.958501            0                     0   11.174432   \n",
       "2  0.025962          0.958501            0                     0    2.769318   \n",
       "3  0.028089          0.958501            0                     0    4.860584   \n",
       "4  0.031491          0.958501            0                     0    3.523450   \n",
       "\n",
       "   Combined Score                                              Genes   cohort  \\\n",
       "0       15.304373  SEC61A1;SEC24A;MAN1A2;UBE4B;DNAJC5G;STT3B;UGGT...  BRCA_5q   \n",
       "1       44.937142                                         RAD50;LIG4  BRCA_5q   \n",
       "2       10.111073                    HERC2;UBR5;UBE4B;BRCA1;BTRC;CBL  BRCA_5q   \n",
       "3       17.363799                                 RAD50;RAD51C;BRCA1  BRCA_5q   \n",
       "4       12.184308                          HMGCS1;ACOX1;EHHADH;PLIN2  BRCA_5q   \n",
       "\n",
       "   number_sampled_in_random  \n",
       "0                         0  \n",
       "1                         0  \n",
       "2                         0  \n",
       "3                         0  \n",
       "4                         0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gsea_summary_df = pd.concat(gsea_dfs,\n",
    "                            axis=0,\n",
    "                            ignore_index=True)\n",
    "gsea_summary_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "      <th>cohort</th>\n",
       "      <th>number_sampled_in_random</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5956</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>Herpes simplex virus 1 infection</td>\n",
       "      <td>89/498</td>\n",
       "      <td>5.323906e-12</td>\n",
       "      <td>1.602496e-09</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.456442</td>\n",
       "      <td>63.766331</td>\n",
       "      <td>ZNF254;ZNF573;IFNA5;ZNF253;ZNF492;ITGB3;TBK1;Z...</td>\n",
       "      <td>BRCA_UCEC_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2029</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>Herpes simplex virus 1 infection</td>\n",
       "      <td>75/498</td>\n",
       "      <td>9.394906e-12</td>\n",
       "      <td>2.780892e-09</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.627076</td>\n",
       "      <td>66.703695</td>\n",
       "      <td>ZNF254;ZNF573;ZNF492;ITGB3;TBK1;ZNF729;ZNF529;...</td>\n",
       "      <td>UCEC_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5957</th>\n",
       "      <td>KEGG_2021_Human</td>\n",
       "      <td>Homologous recombination</td>\n",
       "      <td>12/41</td>\n",
       "      <td>9.141805e-05</td>\n",
       "      <td>1.375842e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.549486</td>\n",
       "      <td>42.310528</td>\n",
       "      <td>POLD3;TOP3B;BRIP1;RAD50;RAD51C;XRCC2;RPA2;ATM;...</td>\n",
       "      <td>BRCA_UCEC_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Somatic Recombination Of Immunoglobulin Genes ...</td>\n",
       "      <td>4/9</td>\n",
       "      <td>7.796097e-06</td>\n",
       "      <td>1.407195e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49.491824</td>\n",
       "      <td>582.117258</td>\n",
       "      <td>MSH6;SANBR;RNF8;LIG4</td>\n",
       "      <td>BRCA_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Isotype Switching (GO:0045190)</td>\n",
       "      <td>4/11</td>\n",
       "      <td>1.990371e-05</td>\n",
       "      <td>1.796310e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>35.347709</td>\n",
       "      <td>382.624963</td>\n",
       "      <td>MSH6;SANBR;RNF8;LIG4</td>\n",
       "      <td>BRCA_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6257</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Nuclear Pore Organization (GO:0006999)</td>\n",
       "      <td>8/14</td>\n",
       "      <td>4.534526e-06</td>\n",
       "      <td>1.810183e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14.642686</td>\n",
       "      <td>180.160530</td>\n",
       "      <td>NDC1;NUP107;SEH1L;NUP133;TPR;NUP153;NUP98;RTN4</td>\n",
       "      <td>BRCA_UCEC_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2325</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>DNA Damage Response (GO:0006974)</td>\n",
       "      <td>49/384</td>\n",
       "      <td>5.569501e-06</td>\n",
       "      <td>2.022286e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.132688</td>\n",
       "      <td>25.801692</td>\n",
       "      <td>TOP2A;WDR48;BRCA1;FMN2;INSL6;NIPBL;KMT5B;MACRO...</td>\n",
       "      <td>UCEC_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6259</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Protein Alpha-1,2-Demannosylation (GO:0036508)</td>\n",
       "      <td>8/16</td>\n",
       "      <td>1.667552e-05</td>\n",
       "      <td>2.218956e-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10.980815</td>\n",
       "      <td>120.806196</td>\n",
       "      <td>EDEM3;RNF139;MAN1A2;EDEM2;MAN1C1;UGGT2;UGGT1;M...</td>\n",
       "      <td>BRCA_UCEC_5q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Gene_set  \\\n",
       "5956             KEGG_2021_Human   \n",
       "2029             KEGG_2021_Human   \n",
       "5957             KEGG_2021_Human   \n",
       "224   GO_Biological_Process_2023   \n",
       "225   GO_Biological_Process_2023   \n",
       "6257  GO_Biological_Process_2023   \n",
       "2325  GO_Biological_Process_2023   \n",
       "6259  GO_Biological_Process_2023   \n",
       "\n",
       "                                                   Term Overlap       P-value  \\\n",
       "5956                   Herpes simplex virus 1 infection  89/498  5.323906e-12   \n",
       "2029                   Herpes simplex virus 1 infection  75/498  9.394906e-12   \n",
       "5957                           Homologous recombination   12/41  9.141805e-05   \n",
       "224   Somatic Recombination Of Immunoglobulin Genes ...     4/9  7.796097e-06   \n",
       "225                      Isotype Switching (GO:0045190)    4/11  1.990371e-05   \n",
       "6257             Nuclear Pore Organization (GO:0006999)    8/14  4.534526e-06   \n",
       "2325                   DNA Damage Response (GO:0006974)  49/384  5.569501e-06   \n",
       "6259     Protein Alpha-1,2-Demannosylation (GO:0036508)    8/16  1.667552e-05   \n",
       "\n",
       "      Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
       "5956      1.602496e-09            0                     0    2.456442   \n",
       "2029      2.780892e-09            0                     0    2.627076   \n",
       "5957      1.375842e-02            0                     0    4.549486   \n",
       "224       1.407195e-02            0                     0   49.491824   \n",
       "225       1.796310e-02            0                     0   35.347709   \n",
       "6257      1.810183e-02            0                     0   14.642686   \n",
       "2325      2.022286e-02            0                     0    2.132688   \n",
       "6259      2.218956e-02            0                     0   10.980815   \n",
       "\n",
       "      Combined Score                                              Genes  \\\n",
       "5956       63.766331  ZNF254;ZNF573;IFNA5;ZNF253;ZNF492;ITGB3;TBK1;Z...   \n",
       "2029       66.703695  ZNF254;ZNF573;ZNF492;ITGB3;TBK1;ZNF729;ZNF529;...   \n",
       "5957       42.310528  POLD3;TOP3B;BRIP1;RAD50;RAD51C;XRCC2;RPA2;ATM;...   \n",
       "224       582.117258                               MSH6;SANBR;RNF8;LIG4   \n",
       "225       382.624963                               MSH6;SANBR;RNF8;LIG4   \n",
       "6257      180.160530     NDC1;NUP107;SEH1L;NUP133;TPR;NUP153;NUP98;RTN4   \n",
       "2325       25.801692  TOP2A;WDR48;BRCA1;FMN2;INSL6;NIPBL;KMT5B;MACRO...   \n",
       "6259      120.806196  EDEM3;RNF139;MAN1A2;EDEM2;MAN1C1;UGGT2;UGGT1;M...   \n",
       "\n",
       "            cohort  number_sampled_in_random  \n",
       "5956  BRCA_UCEC_5q                         0  \n",
       "2029       UCEC_5q                         0  \n",
       "5957  BRCA_UCEC_5q                         0  \n",
       "224        BRCA_5q                         0  \n",
       "225        BRCA_5q                         0  \n",
       "6257  BRCA_UCEC_5q                         0  \n",
       "2325       UCEC_5q                         0  \n",
       "6259  BRCA_UCEC_5q                         0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gsea_summary_df.loc[gsea_summary_df[\"Adjusted P-value\"].sort_values().iloc[:8].index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save combined tables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_summary_df.to_csv(gsea_summary_filename,\n",
    "                       sep=\"\\t\",\n",
    "                       index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "codon_context",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
